### DISCUSSION TITLE
Time Apodization in DFT field monitors

### DISCUSSION THREAD

**Discussion by raksharuia** [NanoComp/meep#2914] [Q&A] [answered]
Hello,

I am performing cavity simulations with a source and using the DFT field monitors to get electric field profiles at resonant frequencies. The source is however quite dominant in the output fields. I want to use "Temporal apodization" at the beginning of the time series (to remove the contribution of the source excitation) before FFT is performed to compute fields at a given frequency. I did not locate this in the 'add_dft_field' method. Is there any way to set this up in MEEP. 

I noticed there was some discussion in the previous issue where someone suggested these features from Lumerical (https://github.com/NanoComp/meep/issues/291), albeit the basic discussion was different.

**Comment by stevengj** [comment]
It's a linear calculation so you typically just need to normalize the output fields by the source.  e.g. do a normalization calculation with the same source in a control structure with no cavity (vacuum? a waveguide? I don't quite know what you are simulating) and then divide the two.

This is the same as how you do transmission calculations, etc.   I don't think apodization has anything to do with it.

Hard to provide more specific advice without knowing precisely what cavity properties you are trying to calculate.

**Reply by raksharuia** [reply]
Thank you for quick reply :-)

I am simulating micropillar cavities (lambda layer sandwiched between top and bottom DBR) with a dipole source at the center. 

The excitation frequency (for source) used for the simulation is near that of resonance mode of the cavity. I want to use the dft monitors to study the resonance mode profiles; the field data could be further used to compute mode sizes - like area and/or diameter. 

The reason I wish to use apodization is to remove the initial segment of time series from the field data to exclude that contribution from the Fourier transform.

Without apodization, an example mode profile (using dft field monitors) of such a system with source would look like
![image](https://github.com/user-attachments/assets/7c4c53b2-c79a-414d-88b3-b94fbc1b31ba)

If time apodization to exclude the source signal were performed before fourier transform, the mode should look like - 
(I generated this second image from a 2D Gaussian fit of the previous curve). I would say this is not a accurate picture of the mode (quantitatively) as the ripples and the sharp peak (effect of the initial source excitation) get incorporated here.
![image](https://github.com/user-attachments/assets/bac37c60-f76e-42d6-9f24-0f20bf96b66d)

Better fit/post-processing algorithms or normalization would be good solutions here. I was curious if these features (like apodization) had been incorporated and/or something else was available to extract resonance mode fields at given frequencies in such a simulation where one needs to explore modes after the initial transients of source excitation.

**Reply by stevengj** [reply, marked as answer]
> The reason I wish to use apodization is to remove the initial segment of time series from the field data to exclude that contribution from the Fourier transform.

You can just run the simulation until the source is turned off, and *then* add the DFT monitor, and then run for some more time.

But if you have a cavity with a strong resonance, and you excite it at the resonant frequency, you don't need a DFT at all — just excite it with a narrow-band source, and then look at a snapshot of the field after the source has finished — everything except the resonant mode should have decayed away.   See, for example, this tutorial: https://meep.readthedocs.io/en/latest/Python_Tutorials/Basics/#modes-of-a-ring-resonator
